pkg/controller/bucket/bucket_controller.go

```diff
@@ -0,0 +1,200 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package bucket
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilversion "k8s.io/apimachinery/pkg/util/version"
+
+	kubeclientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/client-go/util/workqueue"
+
+	"github.com/kubernetes-sigs/container-object-storage-interface-api/apis/objectstorage.k8s.io/v1alpha1"
+	bucketclientset "github.com/kubernetes-sigs/container-object-storage-interface-api/clientset"
+	"github.com/kubernetes-sigs/container-object-storage-interface-api/controller"
+
+	osspec "github.com/kubernetes-sigs/container-object-storage-interface-spec"
+
+	"k8s.io/klog"
+
+	"golang.org/x/time/rate"
+)
+
+// bucketListener manages Bucket objects
+type bucketListener struct {
+	kubeClient        kubeclientset.Interface
+	bucketClient      bucketclientset.Interface
+	provisionerClient osspec.ProvisionerClient
+
+	// The name of the provisioner for which this controller dynamically
+	// provisions buckets.
+	provisionerName string
+	kubeVersion     *utilversion.Version
+}
+
+// NewBucketController returns a controller that manages Bucket objects
+func NewBucketController(provisionerName string, client osspec.ProvisionerClient) (*controller.ObjectStorageController, error) {
+	rateLimit := workqueue.NewMaxOfRateLimiter(
+		workqueue.NewItemExponentialFailureRateLimiter(5*time.Second, 60*time.Minute),
+		&workqueue.BucketRateLimiter{Limiter: rate.NewLimiter(rate.Limit(10), 100)},
+	)
+
+	identity := fmt.Sprintf("object-storage-sidecar-%s", provisionerName)
```

Comment:
I think we would want to standardize the names objectstorage-sidecar-XXXX

Reply:
I'll do that when I push up the main binary bits.